### PR TITLE
Implement optuna.terminator using optuna._gp

### PR DESCRIPTION
**Pull request by y0z**
## Motivation
- Change the GP implementation used in `optuna.terminator` to `optuna._gp` from BoTorch.

## Description of the changes
- Refactoring `optuna.terminator` and related code for using `optuna._gp`.
- Remove files that are no longer needed.

### CODE REVIEW

**Comment by contramundum53** [comment]
image
The result looks good.

```
import optuna
import numpy as np
from optuna.terminator import RegretBoundEvaluator

study = optuna.study.create_study(sampler=optuna.samplers.TPESampler())

def objective(trial):
    x = trial.suggest_float("x", -3, 3)
    y = trial.suggest_float("y", -3, 3)
    return x ** 2 + y ** 2



study.optimize(objective, n_trials=50)
from optuna.visualization import plot_terminator_improvement

plot_terminator_improvement(study, improvement_evaluator=RegretBoundEvaluator(top_trials_ratio=1.0))
```

**Review by contramundum53 — APPROVED** [review]
LGTM!

**Comment by contramundum53** [comment]
@nabenabe0928

**Comment by contramundum53** [comment]
Follow-up tasks:
* Use `kernel_params_cache` to speed up evaluation
* Set `DEFAULT_TOP_TRIALS_RATIO` to 1.0
* Set `beta` to constant

File: optuna/terminator/improvement/evaluator.py [hunk 1]
```diff
@@ -64,53 +87,98 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
...
-
-        processes += [OneToHot()]
-
-        return PreprocessingPipeline(processes)
+        self._log_prior = prior.default_log_prior
+        self._minimum_noise = prior.DEFAULT_MINIMUM_NOISE_VAR
+        self._optimize_n_samples = 2048
+        self._rng = LazyRandomState(seed)
 
     def evaluate(
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:101]: I would code like this to narrow down the variable scopes and enhance the readability.
IMO, the calculation of standardized regret bound should be separated from the main method.

```python
    def _compute_standardized_regret_bound(
        self,
        kernel_params: "gp.KernelParamsTensor",
        gp_search_space: "search_space.SearchSpace",
        normalized_params: np.ndarray,
        standardized_scores: np.ndarray,
    ) -> float:
        """
        In the original paper, f(x) was intended to be minimized, but here we would like to
        maximize f(x), i.e. score_vals. Hence, the following changes happen:
            1. min(ucb) over top trials becomes max(lcb) over top trials, and
            2. min(lcb) over the search space becomes max(ucb) over the search space, and
            3. Regret bound becomes max(ucb) over the search space minus max(lcb) over top trials.
        """
        n_trials, n_params = normalized_params.shape
        sqrt_beta = np.sqrt(_get_beta(n_params=n_params, n_trials=n_trials))
        ucb_acqf_params = acqf.create_acqf_params(
            acqf_type=acqf.AcquisitionFunctionType.UCB,
            kernel_params=kernel_params,
            search_space=gp_search_space,
            X=normalized_params,
            Y=standardized_scores,
            sqrt_beta=sqrt_beta,
        )
        seed = self._rng.rng.randint(np.iinfo(np.int32).max)
        # UCB over the search space. (Original: LCB over the search space. See Change 1 above.)
        max_ucb = np.max(
            acqf.eval_acqf_no_grad(ucb_acqf_params, normalized_params),
            initial=optim.optimize_acqf_sample(ucb_acqf_params, self._optimize_n_samples, seed)[1],
        )
        lcb_acqf_params = acqf.create_acqf_params(
            acqf_type=acqf.AcquisitionFunctionType.UCB,
            kernel_params=kernel_params,
            search_space=gp_search_space,
            X=normalized_params,
            Y=standardized_scores,
            sqrt_beta=-sqrt_beta,
        )
        # LCB over the top trials. (Original: UCB over the top trials. See Change 2 above.)
        max_lcb = np.max(acqf.eval_acqf_no_grad(lcb_acqf_params, normalized_params))
        # max(UCB) - max(LCB). (Original: min(UCB) - min(LCB). See Change 3 above.)
        return max_ucb - max_lcb

    def _get_top_n(
        self, params: np.ndarray, score_vals: np.ndarray
    ) -> tuple[np.ndarray, np.ndarray]:
        assert len(params) == len(score_vals)
        n_trials = len(params)
        top_n = np.clip(int(n_trials * self._top_trials_ratio), self._min_n_trials, n_trials)
        top_n_val = np.partition(score_vals, n_trials - top_n)[n_trials - top_n]
        top_mask = score_vals >= top_n_val
        return params[top_mask], score_vals[top_mask]

    def evaluate(self, trials: List[FrozenTrial], study_direction: StudyDirection) -> float:
        optuna_search_space = intersection_search_space(trials)
        self._validate_input(trials, optuna_search_space)
        complete_trials = [t for t in trials if t.state == TrialState.COMPLETE]

        # _gp module assumes that optimization direction is maximization
        sign = -1 if study_direction == StudyDirection.MINIMIZE else 1
        score_vals = np.array([t.value for t in complete_trials]) * sign
        gp_search_space, normalized_params = search_space.get_search_space_and_normalized_params(
            complete_trials, optuna_search_space
        )

        top_n_normalized_params, top_n_scores = self._get_top_n(normalized_params, score_vals)
        top_n_score_std = max(1e-10, top_n_scores.std())
        top_n_standardized_scores = (top_n_scores - np.mean(top_n_scores)) / top_n_score_std
        kernel_params = gp.fit_kernel_params(
            X=top_n_normalized_params,
            Y=top_n_standardized_scores,
            is_categorical=(gp_search_space.scale_types == search_space.ScaleType.CATEGORICAL),
            log_prior=self._log_prior,
            minimum_noise=self._minimum_noise,
            initial_kernel_params=None,
        )
        standardized_regret = self._compute_standardized_regret_bound(
            kernel_params=kernel_params,
            gp_search_space=gp_search_space,
            normalized_params=top_n_normalized_params,
            standardized_scores=top_n_standardized_scores,
        )
        return standardized_regret * top_n_score_std
```

**Review by nabenabe0928 — COMMENTED** [review]
I confirmed that we get the identical result with my modification as well!

```python
import optuna
import numpy as np
from optuna.terminator import RegretBoundEvaluator

study = optuna.study.create_study(sampler=optuna.samplers.TPESampler(seed=42))

def objective(trial):
    x = trial.suggest_float("x", -3, 3)
    y = trial.suggest_float("y", -3, 3)
    return x ** 2 + y ** 2



study.optimize(objective, n_trials=50)
from optuna.visualization import plot_terminator_improvement

plot_terminator_improvement(study, improvement_evaluator=RegretBoundEvaluator(top_trials_ratio=1.0, seed=42))
```

File: optuna/terminator/improvement/evaluator.py [hunk 2]
```diff
@@ -64,53 +87,98 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
...
+            )
+        ucb_value = standardized_ucb_value * top_n_values_std + top_n_values_mean
+
+        lcb_acqf_params = acqf.create_acqf_params(
+            acqf_type=acqf.AcquisitionFunctionType.UCB,
+            kernel_params=kernel_params,
+            search_space=gp_search_space,
+            X=normalized_top_n_params,
+            Y=standarized_top_n_values,
+            sqrt_beta=-sqrt_beta,
```
[the remark this answers is not in the document]
Comment (y0z) [reply optuna/terminator/improvement/evaluator.py]: > sqrt_beta is supposedly a positive value

You're right. To resolve this, I added LCB.

File: optuna/terminator/improvement/evaluator.py [hunk 3]
```diff
@@ -64,53 +87,98 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
 
     def __init__(
         self,
-        gp: Optional[BaseGaussianProcess] = None,
```
[the remark this answers is not in the document]
Comment (y0z) [reply optuna/terminator/improvement/evaluator.py:67]: I added the reference.

File: optuna/terminator/improvement/evaluator.py [hunk 4]
```diff
@@ -1,31 +1,54 @@
...
+    acqf = _LazyImport("optuna._gp.acqf")
+    prior = _LazyImport("optuna._gp.prior")
+    search_space = _LazyImport("optuna._gp.search_space")
+
 DEFAULT_TOP_TRIALS_RATIO = 0.5
 DEFAULT_MIN_N_TRIALS = 20
 
 
+def _get_beta(n_params: int, n_trials: int, delta: float = 0.1) -> float:
+    beta = 2 * np.log(n_params * n_trials**2 * np.pi**2 / 6 / delta)
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:42]:
```suggestion
    # TODO(nabenabe0928): Check the original implementation to verify.
    # Especially, |D| seems to be the domain size, but not the dimension based on Theorem 1.
    beta = 2 * np.log(n_params * n_trials**2 * np.pi**2 / 6 / delta)
```

File: optuna/terminator/improvement/evaluator.py [hunk 5]
```diff
@@ -60,59 +83,135 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
         min_lcb_n_additional_samples:
             A minimum number of additional samples to estimate the lower confidence bound.
             Default to 2000.
-    """
+
+    For further information about this evaluator, please refer to the following paper:
+
+    - `Automatic termination for hyperparameter optimization <https://proceedings.mlr.press/v188/makarova22a.html>`_
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:89]: nit

```suggestion
    - `Automatic Termination for Hyperparameter Optimization <https://proceedings.mlr.press/v188/makarova22a.html>`_
```

File: optuna/terminator/improvement/evaluator.py [hunk 6]
```diff
@@ -1,31 +1,54 @@
...
 DEFAULT_MIN_N_TRIALS = 20
 
 
+def _get_beta(n_params: int, n_trials: int, delta: float = 0.1) -> float:
+    beta = 2 * np.log(n_params * n_trials**2 * np.pi**2 / 6 / delta)
+
+    # The following div is according to the original paper: "We then further scale it down
+    # by a factor of 5 as defined in the experiments in
+    # `Srinivas et al. (2010) <https://dl.acm.org/doi/10.5555/3104322.3104451>`_"
+    beta /= 5
```
[the remark this answers is not in the document]
Comment (y0z) [reply optuna/terminator/improvement/evaluator.py:47]: This issue will be resolved in another PR.

File: optuna/terminator/improvement/evaluator.py [hunk 7]
```diff
@@ -60,59 +83,135 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
...
+        maximize f(x). Hence, the following changes happen:
+            1. min(ucb) over top trials becomes max(lcb) over top trials, and
+            2. min(lcb) over the search space becomes max(ucb) over the search space, and
+            3. Regret bound becomes max(ucb) over the search space minus max(lcb) over top trials.
+        """
+
+        n_params = len(optuna_search_space)
+
+        # calculate max_ucb
+        beta = np.sqrt(_get_beta(n_params, len(normalized_top_n_params)))
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:134]: It is not `sqrt_beta` anymore.
```suggestion
        beta = _get_beta(n_params, len(normalized_top_n_params))
```

File: optuna/terminator/improvement/evaluator.py [hunk 8]
```diff
@@ -60,59 +85,129 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
...
+        assert len(normalized_params) == len(values)
+        n_trials = len(normalized_params)
+        top_n = np.clip(int(n_trials * self._top_trials_ratio), self._min_n_trials, n_trials)
+        top_n_val = np.partition(values, n_trials - top_n)[n_trials - top_n]
+        top_n_mask = values >= top_n_val
+        return normalized_params[top_n_mask], values[top_n_mask]
+
+    def _compute_standardized_regret_bound(
+        self,
+        optuna_search_space: Dict[str, BaseDistribution],
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:119]: We can use `normalized_top_n_params` instead.
```suggestion
```

File: optuna/terminator/improvement/evaluator.py [hunk 9]
```diff
@@ -60,59 +85,129 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
...
+        maximize f(x). Hence, the following changes happen:
+            1. min(ucb) over top trials becomes max(lcb) over top trials, and
+            2. min(lcb) over the search space becomes max(ucb) over the search space, and
+            3. Regret bound becomes max(ucb) over the search space minus max(lcb) over top trials.
+        """
+
+        n_params = len(optuna_search_space)
+
+        # calculate max_ucb
+        beta = _get_beta(n_params, len(normalized_top_n_params))
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:136]:
```suggestion
        n_trials, n_params = normalized_top_n_params.shape

        # calculate max_ucb
        beta = _get_beta(n_params, n_trials)
```

File: optuna/terminator/improvement/evaluator.py [hunk 10]
```diff
@@ -60,59 +85,129 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
...
+        _, standardized_ucb_value = optim.optimize_acqf_sample(
+            ucb_acqf_params,
+            n_samples=self._optimize_n_samples,
+            seed=self._rng.rng.randint(np.iinfo(np.int32).max),
+        )
+        # UCB over the search space. (Original: LCB over the search space. See Change 1 above.)
+        standardized_ucb_value = max(
+            standardized_ucb_value,
+            float(acqf.eval_acqf_no_grad(ucb_acqf_params, normalized_top_n_params).max().item()),
+        )
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:154]:
```suggestion
        seed = self._rng.rng.randint(np.iinfo(np.int32).max)
        # UCB over the search space. (Original: LCB over the search space. See Change 1 above.)
        max_ucb = max(
            acqf.eval_acqf_no_grad(ucb_acqf_params, normalized_top_n_params).max(),
            default=optim.optimize_acqf_sample(ucb_acqf_params, self._optimize_n_samples, seed)[1],
        )
```

File: optuna/terminator/improvement/evaluator.py [hunk 11]
```diff
@@ -60,59 +85,129 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
...
+            kernel_params=kernel_params,
+            search_space=gp_search_space,
+            X=normalized_top_n_params,
+            Y=standarized_top_n_values,
+            beta=beta,
+        )
+        # LCB over the top trials. (Original: UCB over the top trials. See Change 2 above.)
+        standardized_lcb_value = float(
+            acqf.eval_acqf_no_grad(lcb_acqf_params, normalized_top_n_params).max().item()
+        )
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:168]:
```suggestion
        max_lcb = np.max(acqf.eval_acqf_no_grad(lcb_acqf_params, normalized_top_n_params))
```

File: optuna/terminator/improvement/evaluator.py [hunk 12]
```diff
@@ -60,59 +85,129 @@ class RegretBoundEvaluator(BaseImprovementEvaluator):
...
+        top_n_values_std = max(1e-10, top_n_values.std())
+        standarized_top_n_values = (top_n_values - top_n_values_mean) / top_n_values_std
+
+        kernel_params = gp.fit_kernel_params(
+            X=normalized_top_n_params,
+            Y=standarized_top_n_values,
+            is_categorical=(gp_search_space.scale_types == search_space.ScaleType.CATEGORICAL),
+            log_prior=self._log_prior,
+            minimum_noise=self._minimum_noise,
+            initial_kernel_params=None,
```
Comment (nabenabe0928) [inline optuna/terminator/improvement/evaluator.py:200]:
```suggestion
            # TODO(y0z): Add `kernel_params_cache` to speedup.
            initial_kernel_params=None,
```

File: tests/terminator_tests/test_callback.py [hunk 13]
```diff
@@ -1,3 +1,13 @@
+import sys
+
+import pytest
+
+
+# TODO(contramundum53): Remove this block after torch supports Python 3.12.
+if sys.version_info >= (3, 12):
+    pytest.skip("PyTorch does not support python 3.12.", allow_module_level=True)
+
+
```
Comment (nabenabe0928) [inline tests/terminator_tests/test_callback.py:10]: Please 
```suggestion
```

File: tests/terminator_tests/test_erroreval.py [hunk 14]
```diff
@@ -1,9 +1,16 @@
...
+import sys
 
 import pytest
 
+
+# TODO(y0z): Remove this block after torch supports Python 3.12.
+if sys.version_info >= (3, 12):
+    pytest.skip("PyTorch does not support python 3.12.", allow_module_level=True)
+
+
```
Comment (nabenabe0928) [inline tests/terminator_tests/test_erroreval.py:13]:
```suggestion

import pytest

```

File: tests/terminator_tests/test_terminator.py [hunk 15]
```diff
@@ -1,7 +1,15 @@
...
+import sys
+
 import pytest
 
+
+# TODO(y0z): Remove this block after torch supports Python 3.12.
+if sys.version_info >= (3, 12):
+    pytest.skip("PyTorch does not support python 3.12.", allow_module_level=True)
+
+
```
Comment (nabenabe0928) [inline tests/terminator_tests/test_terminator.py:12]:
```suggestion
import pytest

```

File: tests/visualization_tests/test_terminator_improvement.py [hunk 16]
```diff
@@ -1,11 +1,18 @@
...
 from typing import Callable
 
 import pytest
 
+
+# TODO(y0z): Remove this block after torch supports Python 3.12.
+if sys.version_info >= (3, 12):
+    pytest.skip("PyTorch does not support python 3.12.", allow_module_level=True)
+
+
```
Comment (nabenabe0928) [inline tests/visualization_tests/test_terminator_improvement.py:15]:
```suggestion
from typing import Any
from typing import Callable

import pytest

```

**Review by nabenabe0928 — COMMENTED** [review]
Let's add the version constraints to workflow.yml.

File: optuna/_gp/acqf.py [hunk 17]
```diff
@@ -122,6 +127,9 @@ def eval_acqf(acqf_params: AcquisitionFunctionParams, x: torch.Tensor) -> torch.
     elif acqf_params.acqf_type == AcquisitionFunctionType.UCB:
         assert acqf_params.beta is not None, "beta must be given to UCB."
         return ucb(mean=mean, var=var, beta=acqf_params.beta)
+    elif acqf_params.acqf_type == AcquisitionFunctionType.LCB:
+        assert acqf_params.beta is not None, "beta must be given to LCB."
+        return ucb(mean=mean, var=var, beta=acqf_params.beta)
```
Comment (nabenabe0928) [inline optuna/_gp/acqf.py:132]:
```suggestion
        return lcb(mean=mean, var=var, beta=acqf_params.beta)
```

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]
I was missing a bug until now.

File: optuna/_gp/acqf.py (hunk 17 above)
Comment (y0z) [reply optuna/_gp/acqf.py]: Thank you!

**Review by contramundum53 — APPROVED** [review]
LGTM again!

**Pull request merged**